src/Event/TwoFactorAuthenticationEventSubscriber.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\oe_authentication\Event;
+
+use Drupal\cas\Event\CasPreLoginEvent;
+use Drupal\cas\Service\CasHelper;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Executable\ExecutableManagerInterface;
+use Drupal\Core\Plugin\Context\ContextHandlerInterface;
+use Drupal\Core\Plugin\Context\EntityContext;
+use Drupal\Core\Plugin\ContextAwarePluginInterface;
+use Drupal\Core\Plugin\FilteredPluginManagerInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\Utility\Error;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Event subscriber that handles two-factor authentication conditions.
+ */
+class TwoFactorAuthenticationEventSubscriber implements EventSubscriberInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * Creates a new instance of this class.
+   *
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $configFactory
+   *   The config factory.
+   * @param \Drupal\Core\Executable\ExecutableManagerInterface&\Drupal\Core\Plugin\FilteredPluginManagerInterface $conditionManager
+   *   The condition manager.
+   * @param \Drupal\Core\Plugin\Context\ContextHandlerInterface $contextHandler
+   *   The context handler.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The logger for oe_authentication.
+   * @param \Drupal\cas\Service\CasHelper $casHelper
+   *   The CAS helper service.
+   */
+  public function __construct(
+    protected ConfigFactoryInterface $configFactory,
+    protected ExecutableManagerInterface&FilteredPluginManagerInterface $conditionManager,
+    protected ContextHandlerInterface $contextHandler,
+    protected LoggerInterface $logger,
+    protected CasHelper $casHelper,
+  ) {}
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    return [
+      CasPreLoginEvent::class => 'checkTwoFactorAuthentication',
+    ];
+  }
+
+  /**
+   * Checks if two-factor authentication is required and used for the login.
+   *
+   * @param \Drupal\cas\Event\CasPreLoginEvent $event
+   *   The pre-login event.
+   */
+  public function checkTwoFactorAuthentication(CasPreLoginEvent $event) {
+    // If the login has been executed with a two-factor authentication method,
+    // there is no need for further checks.
+    if ($this->isLoginWithTwoFactorAuthentication($event)) {
+      return;
+    }
+
+    $config = $this->configFactory->get('oe_authentication.settings');
+    // If forced 2FA is disabled, there is no extra validation needed.
+    if (!$config->get('force_2fa')) {
+      return;
+    }
+
+    $conditions_configuration = $config->get('2fa_conditions');
+    // If no conditions are present, the 2FA should have been enforced on EU
+    // Login side, but didn't happen.
+    if (empty($conditions_configuration)) {
+      $event->cancelLogin($this->t('You are required to log in using a two-factor authentication method.'));
```

Comment:
When we cancel login like this, what happens to the user?
Do they get bounced back to eulogin with a harder requirement?
Or do they just go back to Drupal login form and later have to manually choose a stronger authentication method?

I just tried with cas mock server: Here it stays in Drupal, and I have to login again.
Unfortunately I did not find how to choose a different auth method with cas mock server.

Reply:
It works exactly the same, you are notified and you need to go manually to EU login to change the method.
EU Login sends you back instantly if you are already logged in, so we cannot for sure redirect them there.
Sending them back with a forced request for 2FA would work, but it's way beyond the scope of this ticket.

Reply:
Still i cannot test that with cas mock server.
But ok.